lib/encoding.ts

```diff
@@ -270,3 +270,57 @@ class CompactedCMPRule implements AutoConsentCMPRule {
         return this.r[9].map(this._decodeRuleStep.bind(this));
     }
 }
+
+export function deduplicateRules(rules: AutoConsentCMPRule[]) {
+    const rulesBySelector = new Map<string, AutoConsentCMPRule[]>();
+    const dedupedRules: AutoConsentCMPRule[] = [];
+    const dedupKey = (rule: AutoConsentCMPRule) => {
+        if (rule.intermediate || !rule.runContext?.urlPattern) {
+            return null; // do not deduplicate intermediate or generic rules
+        }
+        return JSON.stringify({
+            mainFrame: rule.runContext?.main,
+            frame: rule.runContext?.frame,
+            detectCmp: rule.detectCmp,
+            detectPopup: rule.detectPopup,
+            optOut: rule.optOut,
+            prehideSelectors: rule.prehideSelectors,
+            cosmetic: rule.cosmetic,
+        });
+    };
+    rules.forEach((rule) => {
+        const key = dedupKey(rule);
+        if (key && typeof key === 'string') {
+            const patterns = rulesBySelector.get(key) || [];
+            patterns.push(rule);
+            rulesBySelector.set(key, patterns);
```

Comment:
The names `rulesBySelector` and `patterns` are confusing. Is it a leftover of past implementation?